pkg/oidcclient/login.go

```diff
@@ -803,6 +958,60 @@ func (h *handlerState) initOIDCDiscovery() error {
 		return fmt.Errorf("could not decode response_modes_supported in OIDC discovery from %q: %w", h.issuer, err)
 	}
 	h.useFormPost = slices.Contains(discoveryClaims.ResponseModesSupported, "form_post")
+
+	return h.maybePerformPinnipedSupervisorIDPDiscovery()
+}
+
+func (h *handlerState) maybePerformPinnipedSupervisorIDPDiscovery() error {
+	// If this OIDC IDP is a Pinniped Supervisor, it will have a reference to the IDP discovery document.
+	// Go to that document and retrieve the IDPs.
+	var pinnipedSupervisorClaims idpdiscoveryv1alpha1.OIDCDiscoveryResponse
+	if err := h.provider.Claims(&pinnipedSupervisorClaims); err != nil {
+		return fmt.Errorf("could not decode the Pinniped IDP discovery document URL in OIDC discovery from %q: %w", h.issuer, err)
+	}
+
+	// This is not an error - it just means that this issuer is not a Pinniped Supervisor.
+	// Note that this package can be used with OIDC IDPs other than Pinniped Supervisor.
+	if pinnipedSupervisorClaims.SupervisorDiscovery.PinnipedIDPsEndpoint == "" {
+		return nil
+	}
+	// This check confirms that the issuer is hosting the IDP discovery document, which would always be the case for
+	// Pinniped Supervisor. Since there are checks above to confirm that the issuer uses HTTPS, IDP discovery will
+	// always use HTTPS.
+	if !strings.HasPrefix(pinnipedSupervisorClaims.SupervisorDiscovery.PinnipedIDPsEndpoint, h.issuer) {
+		return fmt.Errorf("the Pinniped IDP discovery document must always be hosted by the issuer: %q", h.issuer)
+	}
```

Comment:
Rather than check that the endpoint is on the same server, wouldn't it be sufficient to just check that it is an https URL? If the user is trusting this issuer, then the user should also be trusting any URLs that the issuer is pointing them to for IDP discovery. It's still important that the URL is https though, because plain http would break that chain of trust (which starts with the original issuer https URL and CA bundle), so the user should not follow plain http URLs.

Reply:
I think the idea is that this client should make sure that Pinniped's IDP discovery is hosted on the same `FederationDomain`. Since this will always be true for a Pinniped Supervisor, it seems the natural thing to check here.